include/spdk/env.h

```diff
@@ -323,6 +335,546 @@ int spdk_pci_addr_fmt(char *bdf, size_t sz, const struct spdk_pci_addr *addr);
+ * disabled
+ */
+int spdk_ring_set_water_mark(struct spdk_ring *sr, unsigned count);
+
+/**
+ * dump the status of the ring on the console
+ */
+void spdk_ring_dump(FILE *f, const struct spdk_ring *sr);
+
+/**
+ * dump the status of all rings on the console
+ */
+void spdk_ring_list_dump(FILE *f);
+
+/**
+ * search a ring from its name
+ */
+struct spdk_ring *spdk_ring_lookup(const char *name);
+
+#define SPDK_MAX_LCORE 128
+
+/**
+ * Return the ID of the execution unit we are running on.
+ * \return
+ *  Logical core ID (in EAL thread) or LCORE_ID_ANY (in non-EAL thread)
+ */
+unsigned
+spdk_lcore_id(void);
+
+/**
+ * Get the id of the master lcore
+ *
+ * \return
+ *   the id of the master lcore
+ */
+unsigned
+spdk_get_master_lcore(void);
+
+/**
+ * Get the ID of the physical socket of the specified lcore
+ *
+ * \param lcore_id
+ *   the targeted lcore, which MUST be between 0 and RTE_MAX_LCORE-1.
+ * \return
+ *   the ID of lcoreid's physical socket
+ */
+unsigned
+spdk_lcore_to_socket_id(unsigned lcore_id);
+
+
+/**
+ * Test if an lcore is enabled.
+ *
+ * \param lcore_id
+ *   The identifier of the lcore, which MUST be between 0 and
+ *   RTE_MAX_LCORE-1.
+ * \return
+ *   True if the given lcore is enabled; false otherwise.
+ */
+int
+spdk_lcore_is_enabled(unsigned lcore_id);
+
+/**
+ * Return the number of execution units (lcores) on the system.
+ *
+ * \return
+ *   the number of execution units (lcores) on the system.
+ */
+unsigned
+spdk_lcore_count(void);
+
+/**
+ * Get the next enabled lcore ID.
+ *
+ * \param i
+ *   The current lcore (reference).
+ * \param skip_master
+ *   If true, do not return the ID of the master lcore.
+ * \param wrap
+ *   If true, go back to 0 when SPDK_MAX_LCORE is reached; otherwise,
+ *   return SPDK_MAX_LCORE.
+ * \return
+ *   The next lcore_id or SPDK_MAX_LCORE if not found.
+ */
+unsigned
+spdk_get_next_lcore(unsigned i, int skip_master, int wrap);
+
+/**
+ * Macro to browse all running lcores.
+ */
+#define SPDK_LCORE_FOREACH(i)						\
+	for (i = spdk_get_next_lcore(-1, 0, 0);				\
+	     i<SPDK_MAX_LCORE;						\
+	     i = spdk_get_next_lcore(i, 0, 0))
+
+/**
+ * Macro to browse all running lcores except the master lcore.
+ */
+#define SPDK_LCORE_FOREACH_SLAVE(i)					\
+	for (i = spdk_get_next_lcore(-1, 1, 0);				\
+	     i<SPDK_MAX_LCORE;						\
+	     i = spdk_get_next_lcore(i, 1, 0))
+
+/**
+ * Wait until a lcore finished its job.
+ */
+int
+spdk_wait_lcore(unsigned slave_id);
+
+/**
+ * Wait until on all the lcores.
+ */
+void
+spdk_mp_wait_lcore(void);
+
+/**
+ * Get the current state of the lcore.
+ */
+int
+spdk_get_lcore_state(unsigned lcore_id);
+
+/**
+ * Send a message to a slave lcore identified by slave_id to call a
+ * function f with argument arg. Once the execution is done, the
+ * remote lcore switch in FINISHED state.
+ */
+int
+spdk_remote_launch(int (*f)(void *), void *arg, unsigned slave_id);
+
+/**
+ * Memcpy
+ */
+void
+spdk_memcpy(void *dst, const void *src, int len);
```

Comment:
We need to abstract rte_memcpy().  If this function doesn't work because is't not inlined, we can work on that.  But using rte_memcpy doesn't work for us.